runner/src/server/plugins/engine/pageControllers/PageControllerBase.ts

```diff
@@ -367,8 +370,19 @@ export class PageControllerBase {
       relevantState = merge(relevantState, newValue);
 
       //By passing our current relevantState to getNextPage, we will check if we can navigate to this next page (including doing any condition checks if applicable)
-      nextPage = nextPage.getNextPage(relevantState);
-      //If a nextPage is returned, we must have taken that route through the form so continue our iteration with the new page
+      const next = nextPage.getNextPage(relevantState);
+      previousPages.length === 3 && previousPages.shift();
+      next?.path && previousPages.push(next.path);
+
+      /**
+       * This prevents an infinite loop
+       */
+      if (previousPages[0] === next?.path) {
+        nextPage = null;
+      } else {
+        //If a nextPage is returned, we must have taken that route through the form so continue our iteration with the new page
+        nextPage = next;
```

Comment:
Would it make more sense to have all this code in a `#preventInfiniteLoop()` function?

That way you can get rid of the comment block.